sympy/calculus/util.py

```diff
@@ -1631,3 +1632,178 @@ def union(self, other):
+
+    Examples
+    ========
+
+    >>> from sympy import Symbol, sin, Interval, S, exp, pi
+    >>> from sympy.calculus.util import argmin_real
+
+    >>> x = Symbol('x')
+    >>> argmin_real(sin(x), x, Interval(-2*pi, pi))
+    FiniteSet(-pi/2)
+
+    >>> argmin_real((x - 1)**2, x, S.Reals)
+    FiniteSet(1)
+
+    >>> argmin_real(exp(x), x, S.Reals)
+    EmptySet
+    """
+    f = _sympify(f)
+    if symbol is None:
+        if len(f.free_symbols) > 1:
+            raise ValueError(filldedent('''
+                Variable for which argmin is to be determined needs to be
+                specified.'''))
+        else:
+            symbol = list(f.free_symbols)[0]
+    if symbol not in f.free_symbols:
+        raise ValueError(filldedent('''
+            argmax of %s cannot be calculated with respect to %s'''%
+            (f, symbol)))
+
+    return _argMaxMin(f, symbol , domain, max = False)
+
+def return_piecewise_solutions(f, x):
+    from sympy.logic.boolalg import And, Or
+
+    solutions = []
+    if type(f) == Piecewise:
+      for arg in f.args:
+          values = []
+          if isinstance(arg[1], (And, Or)):
+              for a in arg[1].args:
+                  a = a.simplify()
+                  values += a.args
+          else:
+              values = arg[1].simplify().args
+          for v in values:
+              if v.is_Number:
+                  solutions.append(v)
+    elif type(f) == Add or type(f) == Mul:
+      for arg in f.args:
+          solutions += return_piecewise_solutions(arg, x)
+
+    return FiniteSet(*solutions)
+
+def _argMaxMin(f, symbol, domain, max):
+    """
+    Helper function for calculuating argmin/argmax
+    """
+    from sympy.solvers.solveset import solveset
+
+    solns = S.EmptySet
+
+    first_deriv = f.diff(symbol)
+    if first_deriv.is_positive:
+        soln1 = FiniteSet(domain.sup)
+        soln2 = FiniteSet(domain.inf)
+        if max:
+            return FiniteSet(list(soln1)[0]) if soln1.is_subset(domain) else S.EmptySet
+        else:
+            return FiniteSet(list(soln2)[0]) if soln2.is_subset(domain) else S.EmptySet
+    elif first_deriv.is_negative:
+        soln1 = FiniteSet(domain.sup)
+        soln2 = FiniteSet(domain.inf)
+        if max:
+            return list(soln2)[0] if soln2.is_subset(domain) else S.EmptySet
+        else:
+            return list(soln1)[0] if soln1.is_subset(domain) else S.EmptySet
+
+    try:
+        frange = function_range(f, symbol, domain)
+        if max:
+            extremum = frange.sup
+        else:
+            extremum = frange.inf
+    except NotImplementedError:
+        raise NotImplementedError(filldedent('''
+            Methods for finding the argument of maxima of %s have not been
+            implemented yet.'''%(f)))
+    # If the supremum or the infimum of the function is not in its range,
+    # then an empty set is returned.
+    if not FiniteSet(extremum).is_subset(frange):
+        return EmptySet()
```

Comment:
The `is_subset` function can return True, False or None (for indeterminate). We should only return the EmptySet if we are sure that extremum is not in frange.

It isn't necessary to create a FiniteSet when we can use `frange.contains(extremum)`. That also can return None (an unevaluated Contains) though.